activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -159,6 +159,11 @@ def begin_transaction(options = {})
 
       def commit_transaction
         inner_transaction = @stack.pop
+
+        unless current_transaction.joinable?
+          inner_transaction.records.each(&:call_before_commit)
+        end
+
         inner_transaction.commit
 
         if current_transaction.joinable?
```

Comment:
Think readability may be improved by combining these separate conditionals, giving us two clear branches: when the outer txn is joinable, we commit the inner and hoist its records up to the outer; when the outer txn is not joinable, we commit the inner wrapped by pre- and post-committing its records.